src/generator/declaration_visitor.ts

```diff
@@ -1,26 +1,57 @@
-import {FunctionDeclaration, FunctionExpression, isIdentifier, LVal, VariableDeclarator} from '@babel/types';
+import {
+    AssignmentExpression,
+    FunctionDeclaration,
+    isArrayExpression,
+    isIdentifier,
+    LVal,
+    VariableDeclarator,
+} from '@babel/types';
 import Visitor from '../visitor';
-type Mapping = Map<string, number>;
+
+type VariableMapping = Map<string, number>;
 
 class DeclarationVisitor extends Visitor {
 
+    private static memoryOffset: number = 0;
+
+    private static registerArray(val: LVal, length: number, map: VariableMapping) {
+        if (!isIdentifier(val)) {
+            throw new Error('LValue is not an identifier');
+        }
+
+        // Skip one memory offset to store the size in it
+        map.set(val.name, ++DeclarationVisitor.memoryOffset * 4);
+        DeclarationVisitor.memoryOffset += length;
+    }
+
     private index: number = 0;
-    private parameters: Mapping = new Map();
-    private variables: Mapping = new Map();
+    private parameters: VariableMapping = new Map();
+    private variables: VariableMapping = new Map();
+    private localArrayPointers: VariableMapping = new Map();
 
-    public run(tree: FunctionDeclaration): [Mapping, Mapping] {
+    public run(tree: FunctionDeclaration): [VariableMapping, VariableMapping, VariableMapping] {
         tree.params.forEach((node) => this.registerDeclaration(node, this.parameters));
 
         this.visit(tree.body);
 
-        return [this.parameters, this.variables];
+        return [this.parameters, this.variables, this.localArrayPointers];
     }
 
     protected visitVariableDeclarator(node: VariableDeclarator) {
+        if (node.init !== null && isArrayExpression(node.init)) {
+            DeclarationVisitor.registerArray(node.id, node.init.elements.length, this.localArrayPointers);
+        }
+
         this.registerDeclaration(node.id, this.variables);
     }
 
-    private registerDeclaration(val: LVal, map: Mapping) {
+    protected visitAssignmentExpression(node: AssignmentExpression) {
+        if (isArrayExpression(node.right)) {
```

Comment:
You could use visitArrayExpression directly here, can't you?
Because an array expression can also be part of a function call e.g. 
```javascript
otherFunction([1, 2, 3]);
```

Reply:
We decided to not allow this logic in our sublanguage.